server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java

```diff
@@ -0,0 +1,174 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.jms;
+
+import static org.apache.james.queue.api.Mails.defaultMail;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.util.stream.IntStream;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.metrics.api.Gauge;
+import org.apache.james.metrics.api.GaugeRegistry;
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.queue.api.MailQueue;
+import org.apache.james.queue.api.MailQueueContract;
+import org.apache.mailet.base.test.FakeMail;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+
+import com.github.fge.lambdas.Throwing;
+
+@ExtendWith(JMSMailQueueMetricExtension.class)
+public interface JMSMailQueueMetricContract extends MailQueueContract {
```

Comment:
This should be global to all mail queues and not just JMS.

Reply:
I find it hard to do an abstract `MailQueueMetricContract ` because I need to recreate the new testee(for only metric testing) which we will inject our mock instance later. Do that stuff at `queue-api` level is require more ugly code like I did in the first commit. If anyone have a better implementation, I'm very happy

Reply:
I think that adding a _GaugeRegistry getGaugeRegistry()_ method in **MailQueue** interface will solve your problem; that way you explain that your MailQueue has to support metrics.

Reply:
Or... Better than exposing MQ internals, add a `default or static` method in the test contract to retrieve it in the implementation...